[SRVCOM-2368] Serverless multitenant istio docs (#68618)
openshift/openshift-docs: modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc

+
--
* Deny all traffic that is not explicitly allowed in `knative-serving` and `knative-eventing`
* Allow traffic from `istio-system` and `knative-serving` to activator
* Allow traffic from `knative-serving` to autoscaler
* Allow health probes for Apache Kafka components in `knative-eventing`
* Allow internal traffic for channel based brokers in `knative-eventing`
--

. Apply the authorization policy configuration:
+
[source,terminal]
----
$ oc apply -f knative-default-authz-policies.yaml
----

. Define which OpenShift projects can communicate with each other. For that, every OpenShift project of a tenant needs:
+
--
* One `AuthorizationPolicy` limiting directly incoming traffic to the tenants OpenShift project
* One `AuthorizationPolicy` limiting incoming traffic via the activator component of OpenShift Serverless that runs in the `knative-serving` OpenShift project
* One `AuthorizationPolicy` allowing Kubernetes to call `PreStopHooks` on Knative Services
--
+
Instead of creating these policies by hand, run the `helm` utility to create the necessary resources for each tenant:
+
.Example configuration for `team alpha`
[source,terminal]
----

> $ helm template oci://quay.io/openshift-knative/knative-istio-authz-onboarding --version 1.31.0 --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml

Reviewer: There will be a productized version of the helm chart. So this probably needs to change to that URL. cc @Kaustubh-pande

Reviewer: Also ideally, the version (1.31.0) is from a parameter, so that it automatically matches the version we ship. Otherwise you'll have to update this on every release.

Reviewer 2: Now that the helm chart was published, I think the right way to install it is:
```
$ helm repo add openshift-helm-charts https://charts.openshift.io/
$ helm template openshift-helm-charts/redhat-knative-istio-authz --version 1.31.0 --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml
```
The first command will be done only once. The second command will be also called later for "team bravo"

The `oci://` stuff was used for developemtn purporses.

Reviewer 2: We use/test it here: https://github.com/openshift-knative/serverless-operator/blob/main/hack/generate/mesh-auth-policies.sh#L29

:_content-type: PROCEDURE
[id="serverless-ossm-traffic-isolation-securing-the-service-mesh_{context}"]
= Securing the {SMProductShortName}

> Authorization Policies and mTLS allow you to secure {SMProductShortName}.

Reviewer 3: ```suggestion
Authorization policies and mTLS allow you to secure {SMProductShortName}.
```
Unless you want it to be "`AuthorizationPolicy` objects".

.Procedure

. Make sure that all OpenShift projects of your tenant are part of the same `ServiceMeshMemberRoll` object as members:
+
[source,yaml]
----
apiVersion: maistra.io/v1
kind: ServiceMeshMemberRoll
metadata:
 name: default
 namespace: istio-system
spec:
 members:
   - knative-serving    # static value, needs to be here, see setup page
   - knative-eventing   # static value, needs to be here, see setup page
   - team-alpha-1       # example OpenShift project that belongs to the team-alpha tenant
   - team-alpha-2       # example OpenShift project that belongs th the team-alpha tenant
   - team-bravo-1       # example OpenShift project that belongs to the team-bravo tenant
   - team-bravo-2       # example OpenShift project that belongs th the team-bravo tenant
----
+

> All OpenShift projects that are part of the mesh must enforce mTLS in strict mode. This forces Istio to only accept connections with a client-certificate present and allows the Service Mesh sidecar to validate the origin using `AuthorizationPolicy`.

Reviewer 3: ```suggestion
All OpenShift projects that are part of the mesh must enforce mTLS in strict mode. This forces Istio to only accept connections with a client-certificate present and allows the Service Mesh sidecar to validate the origin using an `AuthorizationPolicy` object.
```

> . Create the configuration with `AuthorizationPolicies` in the `knative-serving` and the `knative-eventing` namespace:

Reviewer 3: ```suggestion
. Create the configuration with `AuthorizationPolicy` objects in the `knative-serving` and `knative-eventing` namespaces:
```

  namespace: knative-eventing
spec:
  action: ALLOW
  selector:
    matchLabels:
      app.kubernetes.io/component: "kafka-channel-receiver"
  rules:
    - from:
        - source:
            namespaces: [ "knative-eventing" ]
            principals: [ "cluster.local/ns/knative-eventing/sa/kafka-controller" ]
      to:
        - operation:
            methods: [ "GET" ]
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-traffic-to-activator
  namespace: knative-serving
spec:
  selector:
    matchLabels:
      app: activator
  action: ALLOW
  rules:
    - from:
        - source:
            namespaces: [ "knative-serving", "istio-system" ]
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-traffic-to-autoscaler
  namespace: knative-serving
spec:
  selector:
    matchLabels:
      app: autoscaler
  action: ALLOW
  rules:
    - from:
        - source:
            namespaces: [ "knative-serving" ]
----
+
These policies restrict the access rules for the network communication between {ServerlessProductShortName} system components. Specifically, they enforce the following rules:
+
--

> * Deny all traffic that is not explicitly allowed in `knative-serving` and `knative-eventing`

Reviewer 3: ```suggestion
* Deny all traffic that is not explicitly allowed in the `knative-serving` and `knative-eventing` namespaces
```

* Allow traffic from `istio-system` and `knative-serving` to activator
* Allow traffic from `knative-serving` to autoscaler
* Allow health probes for Apache Kafka components in `knative-eventing`

> * Allow internal traffic for channel based brokers in `knative-eventing`

Reviewer 4: "channel-based"

> * Allow internal traffic for channel-based brokers in `knative-eventing`

Reviewer 3: Similarly add "the X namespace" styling where appropriate in these.

--

. Apply the authorization policy configuration:
+
[source,terminal]
----
$ oc apply -f knative-default-authz-policies.yaml
----

> . Define which OpenShift projects can communicate with each other. For that, every OpenShift project of a tenant needs:

Reviewer 3: ```suggestion
. Define which OpenShift projects can communicate with each other. For this communication, every OpenShift project of a tenant requires the following:
```

> . Define which OpenShift projects can communicate with each other. For this communication, every OpenShift project of a tenant requires the following:

Reviewer 3: For a follow-up: A few plain "OpenShift"s here.

+
--
* One `AuthorizationPolicy` limiting directly incoming traffic to the tenants OpenShift project

> * One `AuthorizationPolicy` limiting incoming traffic via the activator component of OpenShift Serverless that runs in the `knative-serving` OpenShift project

Reviewer 4: Use "using" rather than "via" in this instance. See IBM SG.

* One `AuthorizationPolicy` allowing Kubernetes to call `PreStopHooks` on Knative Services
--
+

> Instead of creating these policies by hand, install the `helm` utility and create the necessary resources for each tenant:

Reviewer 4: Use "manually" rather than "by hand".

+
.Installing the `helm` utility
[source,terminal]
----
$ helm repo add openshift-helm-charts https://charts.openshift.io/
----
+
.Creating example configuration for `team alpha`
[source,terminal,subs="attributes+"]
----
$ helm template openshift-helm-charts/redhat-knative-istio-authz --version {ServerlessProductVersion} --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml
----
+
.Creating example configuration for `team bravo`
[source,terminal,subs="attributes+"]
----
$ helm template openshift-helm-charts/redhat-knative-istio-authz --version 1.31.0 --set "name=team-bravo" --set "namespaces={team-bravo-1,team-bravo-2}" > team-bravo.yaml

> ----

Reviewer 3: For a follow-up: Using block titles to describe (what seems like) sequential commands/steps is a bit non-standard. I'd suggest breaking these each into actual substeps.
